app/uk/gov/hmrc/cgtpropertydisposalsfrontend/controllers/returns/uploadsupportingdocs/UploadSupportingEvidenceController.scala

```diff
@@ -0,0 +1,1012 @@
+    redirectToIfNoRequiredPreviousAnswer: Call
+  ): Future[Result] =
+    if (requiredPreviousAnswer(currentAnswers).isDefined) {
+      val backLink = currentAnswers.fold(
+        _ => redirectToIfNoRequiredPreviousAnswer,
+        _ => routes.UploadSupportingEvidenceController.checkYourAnswers()
+      )
+      Ok(page(form(currentAnswers), backLink))
+    } else {
+      Redirect(redirectToIfNoRequiredPreviousAnswer)
+    }
+
+  def doYouWantToUploadSupportingDocuments(): Action[AnyContent] = authenticatedActionWithSessionData.async {
+    implicit request =>
+      withUploadSupportingEvidenceAnswers(request) { (_, _, _, _, answers) =>
+        commonDisplayBehaviour(answers)(
+          form = _.fold(
+            _.doYouWantToUploadSupportingEvidence
+              .fold(doYouWantToUploadSupportingDocumentsForm)(doYouWantToUploadSupportingDocumentsForm.fill),
+            c => doYouWantToUploadSupportingDocumentsForm.fill(c.doYouWantToUploadSupportingEvidence)
+          )
+        )(
+          page = doYouWantToUploadSupportingEvidencePage(_, _)
+        )(
+          requiredPreviousAnswer               = { _ => Some(()) },
+          redirectToIfNoRequiredPreviousAnswer = controllers.returns.routes.TaskListController.taskList()
+        )
+      }
+  }
+
+  @SuppressWarnings(Array("org.wartremover.warts.NonUnitStatements"))
+  def doYouWantToUploadSupportingDocumentsSubmit(): Action[AnyContent] = authenticatedActionWithSessionData.async {
+    implicit request =>
+      withUploadSupportingEvidenceAnswers(request) { (draftReturnId, _, _, fillingOutReturn, answers) =>
+        doYouWantToUploadSupportingDocumentsForm.bindFromRequest.fold(
+          errors =>
+            BadRequest(
+              doYouWantToUploadSupportingEvidencePage(errors, controllers.returns.routes.TaskListController.taskList())
+            ),
+          newDoYouWantToUploadSupportingEvidenceAnswer =>
+            if (newDoYouWantToUploadSupportingEvidenceAnswer) {
+              val updatedAnswers = answers match {
+                case IncompleteUploadSupportingEvidenceAnswers(None, supportingEvidences) =>
+                  IncompleteUploadSupportingEvidenceAnswers(
+                    Some(newDoYouWantToUploadSupportingEvidenceAnswer),
+                    supportingEvidences
+                  )
+                case IncompleteUploadSupportingEvidenceAnswers(
+                    Some(oldDoYouWantToUploadSupportingEvidenceAnswer),
+                    supportingEvidences
+                    ) =>
+                  if (newDoYouWantToUploadSupportingEvidenceAnswer =!= oldDoYouWantToUploadSupportingEvidenceAnswer) {
+                    IncompleteUploadSupportingEvidenceAnswers(
+                      Some(newDoYouWantToUploadSupportingEvidenceAnswer),
+                      List.empty
+                    )
+                  } else {
+                    IncompleteUploadSupportingEvidenceAnswers(
```

Comment:
the intention here would be clearer if you just wrote `answers` here - and it would save a lot of lines of code

Reply:
will be addressed after 6th April